File: src/Instance.cpp
#include "../include/Instance.h"
Instance::Instance(pugi::xml_node& elementNode) noexcept : Element(elementNode, "instance")  {
    this->instanceType = elementNode.name();
    this->libraryItem = elementNode.attribute("libraryItemName").as_string();
}
Instance::Instance(std::string& instanceType, std::string& libraryItem) noexcept {
    this->instanceType = instanceType;
    this->libraryItem = libraryItem;
}
Instance::Instance(const Instance& instance) noexcept : Element(instance) {
    this->instanceType = instance.getInstanceType();
    this->libraryItem = instance.getLibraryItem();
}
std::string Instance::getInstanceType() const noexcept {
    return this->instanceType;
}
std::string Instance::getLibraryItem() const noexcept {
    return this->libraryItem;
}
void Instance::setLibraryItem(const std::string& libraryItem) noexcept {
    if (this->Element::root.attribute("libraryItemName").empty()) this->Element::root.append_attribute("libraryItemName");
    this->Element::root.attribute("libraryItemName").set_value(libraryItem.c_str());
    this->libraryItem = libraryItem;
}

File: CsXFL/Instance.cs
using System.Xml.Linq;
namespace CsXFL;

public class Instance : Element
{
    private static readonly Dictionary<string, string> NodeNameToInstanceType = new Dictionary<string, string>
    {
        {"DOMSymbolInstance", "symbol"},
        {"DOMBitmapInstance", "bitmap"},
        {"DOMVideoInstance", "video"},
        {"DOMCompiledClipInstance", "compiled clip"}
    };
    private static readonly List<string> AcceptableInstanceTypes = new List<string> {"symbol", "bitmap", "embedded video", "linked video", "video",  "compiled clip"};
    private readonly string instanceType;
    private string libraryItemName;
    public string InstanceType { get { return instanceType; } }
    public string LibraryItemName
    {
        get { return libraryItemName; }
        set
        {
            libraryItemName = value;
            root?.SetAttributeValue("libraryItemName", value);
        }
    }

    public Instance(in XElement elementNode) : base(elementNode, "instance")
    {
        instanceType = NodeNameToInstanceType[elementNode.Name.LocalName];
        if (!AcceptableInstanceTypes.Contains(instanceType))
        {
            throw new ArgumentException("Invalid instance type: " + instanceType);
        }
        libraryItemName = (string?)elementNode.Attribute("libraryItemName") ?? string.Empty;
    }
    public Instance(in Instance other) : base(other)
    {
        if (!AcceptableInstanceTypes.Contains(other.instanceType))
        {
            throw new ArgumentException("Invalid instance type: " + other.instanceType);
        }
        instanceType = other.instanceType;
        libraryItemName = other.libraryItemName;
    }
    public Instance(in Item item, string instanceType, string nodeName) : base(item, "instance", nodeName)
    {
        if (!AcceptableInstanceTypes.Contains(instanceType))
        {
            throw new ArgumentException("Invalid instance type: " + instanceType);
        }
        this.instanceType = instanceType;
        libraryItemName = item.Name;
        root!.SetAttributeValue("libraryItemName", libraryItemName);
    }

}